src/examples/learn_quant/scripts/generate_expressions.py

```diff
@@ -0,0 +1,187 @@
+from yaml import dump
+from time import time
+import argparse
+from copy import deepcopy
+import pickle
+import os
+import csv
+
+try:
+    from yaml import CDumper as Dumper
+except ImportError:
+    from yaml import Dumper
+
+from ..quantifier import QuantifierUniverse
+from ..grammar import QuantifierGrammar
+from ..meaning import create_universe
+
+def enumerate_quantifiers(depth, quantifiers_universe: QuantifierUniverse, quantifiers_grammar: QuantifierGrammar):
+
+    expressions_by_meaning = quantifiers_grammar.get_unique_expressions(
+        depth,
+        max_size=2 ** len(quantifiers_universe),
+        unique_key=lambda expr: expr.evaluate(quantifiers_universe),
+        compare_func=lambda e1, e2: len(e1) < len(e2),
+    )
+
+    # filter out the trivial meaning, results in NaNs
+    # iterate over keys, since we need to change the dict itself
+    for meaning in list(expressions_by_meaning.keys()):
+        if len(meaning.referents) == 0:
+            del expressions_by_meaning[meaning]
+    
+    return expressions_by_meaning
+
+def save_quantifiers(expressions_by_meaning, 
+                     out_path="learn_quant/outputs/generated_expressions.yml"):
+    
+    with open(out_path, "w+") as outfile:
+        dump(
+            [
+                expressions_by_meaning[meaning].to_dict()
+                for meaning in expressions_by_meaning
+            ],
+            outfile,
+            Dumper=Dumper,
+        )
+
+def save_generation_run(expressions_by_meaning, master_universe, m_size, x_size, depth):
+    from pathlib import Path
+
+    outpath = Path("learn_quant/outputs/inclusive/" + "M"+str(m_size) + "_" + "X"+str(x_size) + "_" + str("d"+str(depth))) / Path("generated_expressions.yml")
+    Path(outpath).parent.mkdir(parents=True, exist_ok=True)
+
+    print("Saving generated expressions...")
+    save_quantifiers(expressions_by_meaning, outpath)
+
+    # Create a new path for the pickle file
+    pickle_outpath = Path(outpath).parent / "master_universe.pkl"
+
+    # Open the file in write binary mode and dump the object
+    with open(pickle_outpath, 'wb') as f:
+        pickle.dump(master_universe, f)
+
+    print("Master universe has been pickled and saved to", pickle_outpath)
+
+def generate_expressions(quantifiers_grammar, args, save=True):
+    quantifiers_grammar.add_indices_as_primitives(args.m_size, args.weight)
```

Comment:
Ah, so here the grammar method gets called.  Why is this needed, as opposed to the "index" rule from the yml?